patterns/interpreter.md

```diff
@@ -0,0 +1,142 @@
+
+```ignore
+(), (()), ()(), ()(()), ..
+```
+
+Formally a CFG consists of terminal symbols, non-terminal symbols, a start
+(non-terminal) symbol, and production rules.
+
+In the previous example we have two terminal symbols `(` and `)`. We define a
+single non-terminal (at the same time the start) symbol `S` and production rules
+as:
+
+```ignore
+S -> SS
+S -> (S)
+S -> ()
+```
+
+So, in order to generate a string with balanced parentheses we repeatedly apply
+production rules. For example, to derive `()()` we apply the following rules:
+
+```ignore
+S -> SS -> ()(S) -> ()()
+```
+
+## Context Free Grammar for our problem
+
+Recall that our task is translate infix expressions into postfix ones. Let's
+define a context free grammar for a set of infix expressions over `0,...,9, +,-`,
+where:
+
+- terminal symbols: `0,...,9, +,-`
+- non-terminal symbols: `exp, term`
+- start symbol is `exp`
+- and the following are production rules
+
+```ignore
+exp -> exp + term
+exp -> exp - term
+exp -> term
+term -> 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9
+```
+
+In fact this grammar should be further transformed depending on what we are going
+to do with the grammar. For example, we might need to remove left recursion.
+For more details please see [Compilers:Principles,Techniques, and Tools](https://en.wikipedia.org/wiki/Compilers:_Principles,_Techniques,_and_Tools)
+(aka Dragon Book).
+
+## Solution
+
+We simply implement a recursive descent parser. The code panics when an expression
+is syntactically wrong (for exmple `2-34` or `2+5-` are wrong according to the
+grammar definition).
+
+```rust
+struct Interpreter<'a> {
+    it: std::str::Chars<'a>,
+}
+
+impl<'a> Interpreter<'a> {
+
+    pub fn new(infix: &'a String) -> Self {
+        Self { it: infix.chars() }
+    }
+
+    pub fn interpret(&mut self, out: &mut String) {
+        self.exp(out);
+    }
+
+    fn next_char(&mut self) -> Option<char> {
+        self.it.next()
+    }
+
+    fn exp(&mut self, out: &mut String) {
+        self.term(out);
+
+        while let Some(op) = self.next_char() {
+            if op == '+' || op == '-' {
+                self.term(out);
+                out.push(op);
+            } else {
+                panic!("Unexpected symbol '{}'", op);
+            }
+        }
+    }
+
+    fn term(&mut self, out: &mut String) {
+        match self.next_char() {
+            Some(ch) if ch.is_digit(10) => out.push(ch),
+            Some(ch) => panic!("Unexpected symbol '{}'", ch),
+            None => panic!("Unexpected end of string"),
+        }
+    }
+}
+
+pub fn main() {
+    let mut infix = String::from("2+3");
+    let mut postfix = String::new();
+    let mut intr = Interpreter::new(&infix);
+    intr.interpret(&mut postfix);
+    assert_eq!(postfix, "23+");
+
+    infix = String::from("1-2+3-4");
+    postfix = String::new();
+    intr = Interpreter::new(&infix);
+    intr.interpret(&mut postfix);
+    assert_eq!(postfix, "12-3+4-");
+}
+```
+
```

Comment:
For me it's ending a bit abruptly. We could mak a small discussion point here or something?

Reply:
@simonsan  Yeah, I'm going to add. It is not finished. I want to also mention about `macro_rules` how to create DSLs
I mentioned you simply to get feedback on simplicity of the example.

Reply:
Ah ok I get it, yeah, I think that it is much simpler and a good take. 👍🏽